### PR TITLE
WIP2

### PR DESCRIPTION
**Pull request by Faariis**
Added post student in user panel:
- Updated teacher index.js;
- Added newStudent.js file;
- Updated the teacher sidebar.jsx;

Added delete button for deleting students:
- Updated listStudents.js;
- Added a function for deleting a student and a delete button;
- Added link button izmijeni for future use;

Added page for editing pupils in the user panel:
- Added pupils/[id] folders and in them added edit.jsx;
- Added gradePage.js that sends the data;
- Added folder named from and in that folder gradesForm.js that is responsible for displaying the from to the user;
- Added editStudent.js that is used for editing the student information;
- Added studentAcknowledgments.js that handels the post and get of all acknowledgments;

Added routes for the user panel:
- Updated edit.jsx;
- Updated gradesPage.js, added dynamic data;
- Updated gradesForm.js;
- Set regular student as default choice in newStudent.js;

Updated home panel and added routes:
- Updated the index.jsx;
- Created home page;
- In home page added [courseID] folder which contains index.jsx and [studentID].jsx;
- Updated listStudentsHome.js (now it shows an input field where we type the name of the student and see his points in every course);
- Updated listCoursesHome.js now it displays the courses (the name of the course is now a link that takes us to listStudentsPerCourse.js);
- Added new file listStudentsPerCourse.js (it lists all the students for that course and there is a button next to every student that takes us to studentDetail.js);
- Added new file studentDetail.js (it contains a table that lists the student information and points);
- Added sidebarHomeStudent.jsx;
- Added responsive design to some parts;

### CODE REVIEW

**Comment by Faariis** [comment]
@IsmetGlumcevic, poštovanje. Evo otvorio sam novi pull request. Problem sa učitavanjem (koje dugo traje) se nalazi unutar listStudentsPerCourse.js.
Ruta 1. '${Url}api/sec-students/student-list' vraća ovaj JSON response: 
[
    {
        "id": 1,
        "primary_school": "Edhem Mulabdic",
        "name": "Neeekoo",
        "middle_name": null,
        "last_name": "Neko",
        "gender": "M",
        "address": "Nedjeljica",
        "guardian_name": "Neki",
        "phone_number": "+38761123456",
        "guardian_number": "+38761123456",
        "guardian_email": "[guardina@email.com](mailto:guardina@email.com)",
        "email": "[neko@neko.com](mailto:neko@neko.com)",
        "date_of_enrollment": "2024-04-01",
        "special_case": "regular",
        "secondary_shool_id": 1,
        "desired_course_A": "RTiA",
        "acknowledgment": null
    },
    {
        "id": 2,
        "primary_school": "Mesa Selimovic",
        "name": "Taj",
        "middle_name": null,
        "last_name": "Neki",
        "gender": "M",
        "address": "Adresa",
        "guardian_name": "Neki guardian",
        "phone_number": "+38761123457",
        "guardian_number": "+38761123458",
        "guardian_email": "[guardina2@email.com](mailto:guardina2@email.com)",
        "email": "[neko@neko2.com](mailto:neko@neko2.com)",
        "date_of_enrollment": "2024-04-01",
        "special_case": "regular",
        "secondary_shool_id": 1,
        "desired_course_A": "RTiA",
        "acknowledgment": null
    },
    {
        "id": 3,
        "primary_school": "Edhem Mulabdic",
        "name": "Johnn",
        "middle_name": null,
        "last_name": "Doe",
        "gender": "M",
        "address": "Adresa2",
        "guardian_name": "Neki guardian2",
        "phone_number": "+38761123458",
        "guardian_number": "+38761123459",
        "guardian_email": "[guardina3@email.com](mailto:guardina3@email.com)",
        "email": "[neko@neko3.com](mailto:neko@neko3.com)",
        "date_of_enrollment": "2024-04-01",
        "special_case": "regular",
        "secondary_shool_id": 1,
        "desired_course_A": "EL",
        "acknowledgment": null
    },
    {
        "id": 4,
        "primary_school": "Mak Dizdar",
        "name": "Jill",
        "middle_name": null,
        "last_name": "Doe",
        "gender": "M",
        "address": "Adresa3",
        "guardian_name": "Neki guardian3",
        "phone_number": "+38761123459",
        "guardian_number": "+38761123460",
        "guardian_email": "[guardina4@email.com](mailto:guardina4@email.com)",
        "email": "[neko@neko4.com](mailto:neko@neko4.com)",
        "date_of_enrollment": "2024-04-01",
        "special_case": "regular",
        "secondary_shool_id": 1,
        "desired_course_A": "EL",
        "acknowledgment": null
    },
    {
        "id": 5,
        "primary_school": "Skender Kulenovic",
        "name": "Neki4",
        "middle_name": null,
        "last_name": "Neki4",
        "gender": "M",
        "address": "Adresa4",
        "guardian_name": "Neki guardian4",
        "phone_number": "+38761123460",
        "guardian_number": "+38761123461",
        "guardian_email": "[guardina5@email.com](mailto:guardina5@email.com)",
        "email": "[neko@neko5.com](mailto:neko@neko5.com)",
        "date_of_enrollment": "2024-04-01",
        "special_case": "regular",
        "secondary_shool_id": 1,
        "desired_course_A": "EL",
        "acknowledgment": null
    },
    {
        "id": 6,
        "primary_school": "Edhem Mulabdic",
        "name": "Taj5",
        "middle_name": null,
        "last_name": "Neki5",
        "gender": "M",
        "address": "Adresa5",
        "guardian_name": "Neki guardian5",
        "phone_number": "+38761123461",
        "guardian_number": "+38761123462",
        "guardian_email": "[guardina6@email.com](mailto:guardina6@email.com)",
        "email": "[neko@neko6.com](mailto:neko@neko6.com)",
        "date_of_enrollment": "2024-04-01",
        "special_case": "regular",
        "secondary_shool_id": 1,
        "desired_course_A": "RTiA",
        "acknowledgment": null
    },
    {
        "id": 7,
        "primary_school": "Mak Dizdar",
        "name": "Taj6",
        "middle_name": null,
        "last_name": "Neki6",
        "gender": "M",
        "address": "Adresa6",
        "guardian_name": "Neki guardian6",
        "phone_number": "+38761123462",
        "guardian_number": "+38761123463",
        "guardian_email": "[guardina7@email.com](mailto:guardina7@email.com)",
        "email": "[neko@neko7.com](mailto:neko@neko7.com)",
        "date_of_enrollment": "2024-04-01",
        "special_case": "regular",
        "secondary_shool_id": 1,
        "desired_course_A": "RTiA",
        "acknowledgment": null
    },
    {
        "id": 8,
        "primary_school": "Skender Kulenovic",
        "name": "Taj7",
        "middle_name": null,
        "last_name": "Neki7",
        "gender": "M",
        "address": "Adresa7",
        "guardian_name": "Neki guardian7",
        "phone_number": "+38761123463",
        "guardian_number": "+38761123464",
        "guardian_email": "[guardina8@email.com](mailto:guardina8@email.com)",
        "email": "[neko@neko8.com](mailto:neko@neko8.com)",
        "date_of_enrollment": "2024-04-01",
        "special_case": "regular",
        "secondary_shool_id": 1,
        "desired_course_A": "RTiA",
        "acknowledgment": null
    },
    {
        "id": 9,
        "primary_school": "Mak Dizdar",
        "name": "Taj8",
        "middle_name": null,
        "last_name": "Neki8",
        "gender": "M",
        "address": "Adresa8",
        "guardian_name": "Neki guardian8",
        "phone_number": "+38761123463",
        "guardian_number": "+38761123464",
        "guardian_email": "[guardina8@email.com](mailto:guardina8@email.com)",
        "email": "[neko@neko8.com](mailto:neko@neko8.com)",
        "date_of_enrollment": "2024-04-01",
        "special_case": "regular",
        "secondary_shool_id": 1,
        "desired_course_A": "EL",
        "acknowledgment": null
    },
    {
        "id": 23,
        "primary_school": "mak",
        "name": "hasoo",
        "middle_name": null,
        "last_name": "mujić",
        "gender": "male",
        "address": "Crkvice",
        "guardian_name": "Hasoo",
        "phone_number": "4445454548",
        "guardian_number": "9204863018",
        "guardian_email": "[hasozds@test.com](mailto:hasozds@test.com)",
        "email": "[haso1zs@test.com](mailto:haso1zs@test.com)",
        "date_of_enrollment": "2024-04-09",
        "special_case": "regular",
        "secondary_shool_id": 1,
        "desired_course_A": "EL",
        "acknowledgment": null
    },
    {
        "id": 24,
        "primary_school": "Mak Dizdar",
        "name": "mujoo",
        "middle_name": null,
        "last_name": "mujić",
        "gender": "male",
        "address": "Crkvice",
        "guardian_name": "Hasoo",
        "phone_number": "44454545485",
        "guardian_number": "92048630186",
        "guardian_email": "[mujo4@test.com](mailto:mujo4@test.com)",
        "email": "[mujo5@test.com](mailto:mujo5@test.com)",
        "date_of_enrollment": "2024-04-15",
        "special_case": "others",
        "secondary_shool_id": 1,
        "desired_course_A": "GT",
        "acknowledgment": null
    }
]

Ruta 2. '${Url}/api/sec-students/student/${studentId}/points-summary/' vraća ovaj JSON response:
[
    {
        "course": "RTIA",
        "total_points": 109.54
    },
    {
        "course": "ELEL",
        "total_points": 109.54
    },
    {
        "course": "ELEN",
        "total_points": 109.54
    },
    {
        "course": "MAST",
        "total_points": 110.54
    },
    {
        "course": "MASTK",
        "total_points": 110.54
    },
    {
        "course": "ST",
        "total_points": 109.54
    },
    {
        "course": "AT",
        "total_points": 109.54
    },
    {
        "course": "METT",
        "total_points": 109.54
    },
    {
        "course": "GT",
        "total_points": 109.54
    }
]

File: src/components/home/listStudentsPerCourse.js
Comment (IsmetGlumcevic) [inline src/components/home/listStudentsPerCourse.js:1]: Ovaj dio se treba refakotrisat i na backendu, jer se salje api poziv da se dobiju svi studenti a onda se za svakog studenta salje novi api poziv da se dobiju poeni. Sto vise studenata bude to ce duze trajati dobijanje konacnih podataka.

Isto tako moze se pokusati prebaciti ovaj dio koda u pages/home/[courseID]/index.jsx gdje se poziva ta komponenta i staviti svu logiku u getStaticProps funkciju i staviti revalidate 1min ili drugi interval u zavisnosti koliko se cesto mjenjaju podaci. Ova chache ce raditi ako je nastimano na serveru da radi sa next cache.

Comment (IsmetGlumcevic) [reply src/components/home/listStudentsPerCourse.js:1]: Pokusaj u jedan api poziv prosljediti sve id-ove i vratiti sve poene u jednom responsu.

Comment (IsmetGlumcevic) [reply src/components/home/listStudentsPerCourse.js:1]: Evo jedan prijedlog za refactor da malo poboljsa perfomance

`import { useEffect, useState } from "react";
import { useSession } from "next-auth/react";
import Url from "../../../constants";
import { useRouter } from "next/router";
import { ChevronLeftIcon } from "@heroicons/react/24/outline";
import Link from "next/link";
import StudentDetails from "./studentDetails";  // This is imported but not used in your example. Consider using or removing it.

const useStudents = (courseId, token) => {
    const [students, setStudents] = useState([]);

    useEffect(() => {
        if (!token) return;  // Early return if no token is available

        const getStudents = async () => {
            try {
                const response = await fetch(`${Url}api/sec-students/student-list`, {
                    method: 'GET',
                    headers: { 'Authorization': `Bearer ${token}` }
                });
                const studentsData = await response.json();

                const studentsWithTotalPoints = await Promise.all(
                    studentsData.map(async (student) => {
                        const totalPoints = await getTotalPoints(student.id, student.desired_course_A, token);
                        return { ...student, totalPoints };
                    })
                );

                const sortedStudents = studentsWithTotalPoints.sort((a, b) => b.totalPoints - a.totalPoints);
                setStudents(sortedStudents.filter(student => student.desired_course_A === courseId));
            } catch (error) {
                console.error('Failed to fetch students:', error);
                // Optionally update UI to show an error message
            }
        };

        getStudents();
    }, [courseId, token]);

    return students;
};

const getTotalPoints = async (studentId, courseName, token) => {
    try {
        const response = await fetch(`${Url}api/sec-students/student/${studentId}/points-summary/`, {
            method: 'GET',
            headers: { 'Authorization': `Bearer ${token}` }
        });
        const pointsData = await response.json();
        const coursePoints = pointsData.find(item => item.course.toUpperCase() === courseName.toUpperCase());
        return coursePoints ? coursePoints.total_points : 0;
    } catch (error) {
        console.error('Failed to fetch points:', error);
        return 0;
    }
};

const ListStudentsPerCourse = ({ courseId }) => {
    const router = useRouter();
    const { data: session } = useSession();
    const students = useStudents(courseId, session?.user?.token);

    const navigateBack = () => router.push('/');

    return (
        <div className="flex flex-col">
            <div className="flex items-center justify-between">
                <h1 className="ml-3 text-2xl font-semibold">Lista učenika {courseId}</h1>
                <button className="ml-3 mb-2 flex items-center px-2 py-1 border-none" onClick={navigateBack}>
                    <ChevronLeftIcon className="w-6 h-6 mr-1" />
                </button>
            </div>
            {students.length > 0 && (
                <div className="border-t border-gray-200 px-4 py-5 sm:p-0">
                    <dl className="sm:divide-y sm:divide-gray-200">
                        {students.map((item, index) => (
                            <div key={index} className="py-4 sm:grid sm:grid-cols-3 sm:gap-4 sm:py-5 sm:px-6">
                                <dt className="text-sm font-medium text-gray-500 capitalize">{index + 1}. {item.name} {item.last_name}{(item.special_case === 'invalid' || item.special_case === 'others') && '*'}</dt>
                                <dd className="mt-1 text-sm text-gray-900 sm:col-span-1 sm:mt-0">Bodovi: {item.totalPoints}</dd>
                                <Link href={`/home/${courseId}/${item.id}`} passHref>
                                    <div className="mt-1 text-sm text-black bg-gray-300 sm:col-span-1 px-4 py-2 hover:bg-gray-400 rounded-md flex items-center justify-center shadow-lg"
                                        style={{ alignSelf: "center", width: "100px" }}>Detalji</div>
                                </Link>
                            </div>
                        ))}
                    </dl>
                </div>
            )}
        </div>
    );
};

export default ListStudentsPerCourse;
`

**Pull request closed without merging**